node/trace/agent.js

```diff
@@ -0,0 +1,167 @@
+// Copyright (c) 2015 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+if (!process.addAsyncListener) {
+    require('async-listener');
+}
+
+var DebugLogtron = require('debug-logtron');
+
+var Span = require('./span');
+
+module.exports = Agent;
+
+function Agent () {
+    if (!(this instanceof Agent)) {
+        return new Agent();
+    }
+    var self = this;
+
+    self.currentSpan = null;
+
+    self.asyncListener = process.createAsyncListener({
+        create: function () {
+            // Return the storage that should be scoped to the async operation
+            // that was just created
+            return self.currentSpan;
+        },
+
+        before: function (context, storage) {
+            if (!self.currentSpan) {
+                self.currentSpan = storage;
+            }
+        },
+
+        after: function () {
+            self.currentSpan = null;
+        },
+
+        error: function () {
+            self.currentSpan = null;
+        }
+    });
+
+    process.addAsyncListener(self.asyncListener);
+
+    self.logger = DebugLogtron('tchannelTrace');
+}
+
+Agent._instance = null;
+
+// Tracing agent is a singleton
+Agent.getInstance = function () {
```

Comment:
This is not how singletons work.

Because of duplication between multiple copies & node_modules you will need real global variables.